libcudacxx/include/cuda/__iterator/transform_output_iterator.h

```diff
@@ -0,0 +1,456 @@
+  [[nodiscard]] _CCCL_API constexpr const _Iter& base() const& noexcept
+  {
+    return __current_;
+  }
+
+  //! @brief Extracts the iterator stored in this \p transform_output_iterator
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _CCCL_API constexpr _Iter base() && noexcept(_CUDA_VSTD::is_nothrow_move_constructible_v<_Iter>)
+  {
+    return _CUDA_VSTD::move(__current_);
+  }
+
+  //! @brief Returns a proxy that transforms the input upon assignment
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _CCCL_API constexpr auto operator*() const noexcept(_CUDA_VSTD::is_nothrow_copy_constructible_v<_Iter>)
+  {
+    return __transform_output_proxy{__current_, const_cast<_Fn&>(*__func_)};
+  }
+
+  //! @brief Returns a proxy that transforms the input upon assignment
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _CCCL_API constexpr auto operator*() noexcept(_CUDA_VSTD::is_nothrow_copy_constructible_v<_Iter>)
+  {
+    return __transform_output_proxy{__current_, *__func_};
+  }
+
+  //! @brief Returns a proxy that transforms the input upon assignment storing the current iterator advanced by \p __n
+  //! @param __n The additional offset
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_TEMPLATE(class _Iter2 = _Iter)
+  _CCCL_REQUIRES(_CUDA_VSTD::__iter_can_subscript<_Iter2>)
+  [[nodiscard]] _CCCL_API constexpr auto operator[](difference_type __n) const
+    noexcept(_CUDA_VSTD::is_nothrow_copy_constructible_v<_Iter2> && noexcept(__current_ + __n))
+  {
+    return __transform_output_proxy{__current_ + __n, const_cast<_Fn&>(*__func_)};
+  }
+
+  //! @brief Returns a proxy that transforms the input upon assignment storing the current iterator advanced by \p __n
+  //! @param __n The additional offset
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_TEMPLATE(class _Iter2 = _Iter)
+  _CCCL_REQUIRES(_CUDA_VSTD::__iter_can_subscript<_Iter2>)
+  [[nodiscard]] _CCCL_API constexpr auto operator[](difference_type __n) noexcept(
+    _CUDA_VSTD::is_nothrow_copy_constructible_v<_Iter2> && noexcept(__current_ + __n))
+  {
+    return __transform_output_proxy{__current_ + __n, const_cast<_Fn&>(*__func_)};
+  }
+
+  //! @brief Increments the stored iterator
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_API constexpr transform_output_iterator& operator++() noexcept(noexcept(++__current_))
+  {
+    ++__current_;
+    return *this;
+  }
+
+  //! @brief Increments the stored iterator
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_API constexpr auto operator++(int) noexcept(noexcept(++__current_))
+  {
+    if constexpr (_CUDA_VSTD::forward_iterator<_Iter> || _CUDA_VSTD::output_iterator<_Iter, value_type>)
+    {
+      auto __tmp = *this;
+      ++*this;
+      return __tmp;
+    }
+    else
+    {
+      ++__current_;
+    }
+  }
+
+  //! @brief Decrements the stored iterator
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_TEMPLATE(class _Iter2 = _Iter)
+  _CCCL_REQUIRES(_CUDA_VSTD::__iter_can_decrement<_Iter2>)
+  _CCCL_API constexpr transform_output_iterator& operator--() noexcept(noexcept(--__current_))
```

Comment:
@bernhardmgruber sorry, it seems I squashed the commit changing this.

We are now checking whether the iterator supports a given operation rather than just checking for `bidirectional_iterator`

That will allow this to work effectively with other output iterators, which do not satisfy `random_access_iterator` but support the random access operations

Reply:
Understood, thx!